### ISSUE TITLE
Warn on unnecessary coercions

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang/rust#56588]
Here the `as *const i32` coercion is unnecessary, but no warning is emitted.

```rust
fn foo(_x: *const i32) {}

pub fn bar(x: &mut i32) {
    foo(x as *const i32)
}

pub fn baz(x: &mut i32) {
    foo(x)
}
```

cc @oli-obk maybe this would be a better fit for clippy ?

**Comment by oli-obk** [comment]
Jup, I'd put it into clippy

**Comment by gnzlbg** [comment]
I've opened https://github.com/rust-lang/rust-clippy/issues/3526
